pkg/forth/forth.go

```diff
@@ -0,0 +1,274 @@
+// Some of the use of error may seem a bit weird but the creation of this package predates the
+// creation of the error type (it was still an os thing back then).
+package forth
+
+import (
+	"errors"
+	"os"
+	"runtime"
+	"strconv"
+	"strings"
+)
+
+type forthop func(f Forth)
+
+type forthstack struct {
+	stack []string
+}
+
+var opmap = map[string]forthop{
+	"+":        plus,
+	"-":        sub,
+	"*":        times,
+	"/":        div,
+	"%":        mod,
+	"swap":     swap,
+	"ifelse":   ifelse,
+	"hostname": hostname,
+	"hostbase": hostbase,
+	"strcat":   strcat,
+	"roundup":  roundup,
+	"dup":      dup,
+}
+
+// Forth is an interface used by the package. The interface
+// requires definition of Push, Pop, Length, Empty (convenience function
+// meaning Length is 0), Newop (insert a new or replacement operator),
+// and Reset (clear the stack, mainly diagnostic)
+type Forth interface {
+	Push(string)
+	Pop() string
+	Length() int
+	Empty() bool
+	Newop(string, forthop)
+	Reset()
+	Stack() []string
+}
+
+// New creates a new stack
+func New() Forth {
+	f := new(forthstack)
+	return f
+}
+
+// Newop creates a new operation. We considered having
+// an opmap per stack but don't feel the package requires it
+func (f *forthstack) Newop(n string, op forthop) {
+	opmap[n] = op
+}
+
+func Ops() map[string]forthop {
+	return opmap
+}
+
+// Reset resets the stack to empty
+func (f *forthstack) Reset() {
+	f.stack = f.stack[0:0]
+}
+
+// Return the stack as a []string
+func (f *forthstack) Stack() []string {
+	return f.stack
+}
+
+// Push pushes the string on the stack.
+func (f *forthstack) Push(s string) {
+	f.stack = append(f.stack, s)
+	//fmt.Printf("push: %v: stack: %v\n", s, f.stack)
+}
+
+// Pop pops the stack. If the stack is Empty Pop will panic.
+// Eval recovers() the panic.
+func (f *forthstack) Pop() (ret string) {
+
+	if len(f.stack) < 1 {
+		panic(errors.New("Empty stack"))
+	}
+	ret = f.stack[len(f.stack)-1]
+	f.stack = f.stack[0 : len(f.stack)-1]
+	//fmt.Printf("Pop: %v stack %v\n", ret, f.stack)
+	return ret
+}
+
+// Length returns the stack length.
+func (f *forthstack) Length() int {
+	return len(f.stack)
+}
+
+// Empty is a convenience function synonymous with Length == 0
+func (f *forthstack) Empty() bool {
+	return len(f.stack) == 0
+}
+
+// errRecover converts panics to errstr iff it is an os.Error, panics
+// otherwise.
+func errRecover(errp *error) {
+	e := recover()
+	if e != nil {
+		if _, ok := e.(runtime.Error); ok {
+			panic(e)
+		}
+		*errp = e.(error)
+	}
+}
+
+/* iEval takes a Forth and strings and splits the string on space
+ * characters, pushing each element on the stack or invoking the
+ * operator if it is found in the opmap.
+ */
+func iEval(f Forth, s string) {
+	// TODO: create a separator list based on isspace and all the
+	// non alpha numberic characters in the opmap.
+	for _, val := range strings.Fields(s) {
+		//fmt.Printf("eval %s stack %v", val, f.Stack())
+		fun := opmap[val]
+		if fun != nil {
```

Comment:
How would you push an arbitrary string ("roundup" for example) onto the stack given that it is already a reserved op? Is there something easier than "r oundup strcat"?

Reply:
I don't think you can, unless there's a quote operator. We can add one if this becomes an issue. It never did in usage. But I could take ` if the first character of a word as a quote operation.